### PR TITLE
Add api for exposing and changing a document's color space

### LINKED ISSUE #531: Add support for Document's color profile
A Document can have a different color profiles (sRGB, P3, auto). We need to expose it in the API.

### PR DESCRIPTION
**Pull request by jedrichards**
Closes #531

### CODE REVIEW

**Comment by jedrichards** [comment]
Need to add docs and tests, but am I on the right track?

**Comment by bohemian-coding-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>This pull request may need a CHANGELOG entry.</td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>

  `Source/dom/models/Document.js` changed, but not:
  - 📚 its docs
- 🧪 its tests

That's OK as long as you're refactoring.
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 1d73769b99363c361de762197d964c8b77f907f9
</p>

**Comment by mathieudutour** [comment]
That's the right track. When you have an enum, I usually have this pattern: https://github.com/BohemianCoding/SketchAPI/blob/develop/Source/dom/style/Fill.js#L118-L133 to return a human readable string instead of a number

**Comment by jedrichards** [comment]
@mathieudutour Ready for another look. There is a difference between assigning and converting the color space, so it made sense to me to add a dedicated method `changeColorSpace` that allowed the user to choose between the two behaviours, and avoid using a setter and making the `colorSpace` prop read only. See what you think ...

File: Source/dom/models/Document.js [hunk 1]
```diff
@@ -389,6 +421,8 @@ if (typeof MSDocument !== 'undefined') {
 }
 
 Document.SaveMode = SaveModeType
+Document.ColorSpace = ColorSpace
+Document.ColorSpaceMap = ColorSpaceMap
```
Comment (mathieudutour) [inline Source/dom/models/Document.js]: no need to expose the map

**Comment by jedrichards** [comment]
@mathieudutour Yep, I down with it! :smile:

File: docs/api/Document.md [hunk 2]
```diff
@@ -221,6 +222,29 @@ document.close()
...
+
+// Pass true as an optional second argument
+// to convert instead of assign
+document.changeColorSpace(ColorSpace.P3, true)
+log(document.colorSpace === ColorSpace.P3) // true
+
+// Alternatively, use the property setter (the behaviour
+// here is to always assign the color space)
+document.colorSpace = ColorSpace.P3
+log(document.colorSpace === ColorSpace.P3) // true
```
Comment (mathieudutour) [inline docs/api/Document.md]: let's use `console.log`

File: Source/dom/models/Document.js [hunk 3]
```diff
@@ -374,6 +386,26 @@ export class Document extends WrappedObject {
 
     msdocument.close()
   }
+
+  changeColorSpace(colorSpace, convert = false) {
```
Comment (mathieudutour) [inline Source/dom/models/Document.js]: I'm wondering if it should be `convertColorSpace` and `convert` should just be `true` :thinking:

Is there a case where you'd use this instead of the `colorSpace` setter?

File: docs/api/Document.md (hunk 2 above)
Comment (jedrichards) [reply docs/api/Document.md]: 👍

File: Source/dom/models/Document.js (hunk 3 above)
Comment (jedrichards) [reply Source/dom/models/Document.js]: I was thinking it could be awkward to have two different apis for assigning vs. converting. By having assigning and converting available via the same method you can switch the behaviour with a one liner:

```
doc.changeColorSpace(ColorSpace.P3, shouldConvert)
```

Rather than

```
if (shouldConvert) {
  doc.convertColorSpace(ColorSpace.P3)
} else {
  doc.colorSpace = ColorSpace.P3
}
```

🤔

Comment (mathieudutour) [reply Source/dom/models/Document.js]: is there a use-case where `shouldConvert` would be dynamic? Tbh I'm still a little be confused why you'd not want to always convert?

Comment (jedrichards) [reply Source/dom/models/Document.js]: I chatted to the Mac devs a bit about this. It sounds like converting is a potentially destructive act when switching between managed color profiles (it walks the document and mutates all the color values), but assigning just sets the property. So in a sense the sort of switch you want to do, assigning vs converting, may depend on the current color profile value ... hence my thinking of adding the flag to the method call.

Basically reproducing some of the similar flow available by the UI:

image

Beyond that, from the point of view of a designer using Sketch, the pros and cons of assigning vs converting are currently beyond my understanding ...

Comment (mathieudutour) [reply Source/dom/models/Document.js]: Ah, maybe ¯\_(ツ)_/¯. Let's keep it that way then

Comment (jedrichards) [reply Source/dom/models/Document.js]: I suppose to be super explicit we could have an api like,

```
document.colorProfile // getter
document.colorProfile = 'P3' /// Assigning setter
document.assignColorProfile('P3')
document.convertToColorProfile('P3')
```

🤔
